nym-vpn-core/crates/nym-authenticator-client/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+        authenticator_address: Recipient,
+    ) -> Result<u64> {
+        let (request, request_id) = match message {
+            ClientMessage::Initial(init_message) => {
+                AuthenticatorRequest::new_initial_request(init_message, self.nym_address)
+            }
+            ClientMessage::Final(gateway_client) => {
+                AuthenticatorRequest::new_final_request(gateway_client, self.nym_address)
+            }
+        };
+        debug!("Sent connect request with version v{}", request.version);
+
+        self.mixnet_sender
+            .send(nym_sdk::mixnet::InputMessage::new_regular(
+                authenticator_address,
+                request.to_bytes().unwrap(),
+                TransmissionLane::General,
+                None,
+            ))
+            .await?;
+
+        Ok(request_id)
+    }
+
+    async fn listen_for_connect_response(&self, request_id: u64) -> Result<AuthenticatorResponse> {
+        // Connecting is basically synchronous from the perspective of the mixnet client, so it's safe
+        // to just grab ahold of the mutex and keep it until we get the response.
+        let mut mixnet_client_handle = self.mixnet_client.lock().await;
+        let mixnet_client = mixnet_client_handle.as_mut().unwrap();
+
+        let timeout = tokio::time::sleep(Duration::from_secs(5));
+        tokio::pin!(timeout);
+
+        loop {
+            tokio::select! {
+                _ = &mut timeout => {
+                    error!("Timed out waiting for reply to connect request");
+                    return Err(Error::TimeoutWaitingForConnectResponse);
+                }
+                msgs = mixnet_client.wait_for_messages() => match msgs {
+                    None => {
+                        return Err(Error::NoMixnetMessagesReceived);
+                    }
+                    Some(msgs) => {
+                        for msg in msgs {
+                            // Confirm that the version is correct
+                            check_auth_message_version(&msg)?;
+
+                            // Then we deserialize the message
+                            debug!("AuthClient: got message while waiting for connect response");
+                            let Ok(response) = AuthenticatorResponse::from_reconstructed_message(&msg) else {
+                                // This is ok, it's likely just one of our self-pings
+                                debug!("Failed to deserialize reconstructed message");
+                                continue;
+                            };
+
+                            if response.id() == Some(request_id) {
+                                debug!("Got response with matching id");
+                                return Ok(response);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn check_auth_message_version(message: &ReconstructedMessage) -> Result<()> {
```

Comment:
I'd rather have this a proper struct (the whole message), but we can leave it for a future iteration